FarihaIS/VanRoomies
# feat(be-chat): add chat functionality to the backend with socket.io

# Welcome to VanRoomies! 👋✈️

Fixes: #43

## Description of the change:
### Socket.io
This change installs a socket.io transport layer on top of our https server. To test, we need a socket.io client that can correctly talk with the server.

### Client
In the `myclient` directory, there exists a JS client that can talk to the server via the socket.io layer. Start it with `npm start`. Then enter an existing `userId` after the prompt.

### Details
After connecting with the socket, we want to send a private message to a user.

To do this we must emit an event through the socket, using the `private message` event. We supply a userId to the socket, and a message, and the server on the other side of the socket then redirects the message to the correct receiver. In other words, the server acts as a proxy and is able to handle which messages go to which user. The server listens on the `private message` event for these messages. This means that all clients must be also listening on the `private message` event to be able to receive messages.

### Message Storage
On the chat message storage side, we use a MongoDB collection to store messages and conversations. The `Conversation` schema is made up of `users` and a `messages` array. Users should only be an array of two `UserId`s, and the messages array is an array that holds `Message` objects of the form:
```js
{
    sender: UserId,
    message: String,
    timestamp: Date.now()
}
```
*`Date.now()` returns the number of milliseconds since epoch: [more info](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/now).*

Every time a message is sent to a socket, we also send a database query to store the message in a `Conversation` between the two users. As long as the user can access the server and the server database, this ensures messages are not lost in-flight if the recipient is offline.

This will require the client to pull all messages from the database initially, and then update the chat in real-time as messages come from the socket.io socket.

## How to manually test:
1. (Add two users to database if you haven't already)
1. On server: `npm run dev`
2. `cd myclient` for client.
3. On client: `npm install` && `npm start`
4. Add another terminal and `npm start` again. (for two clients)
5. On client1: enter a userId of one user on your database.
6. On client2: enter a userId of the other user.
7. Send messages to the other users using their userId after the prompt.
Messages should show up in each client's terminal. 
Also check the database for `conversation` entries.
